### PR TITLE
gh-103295: expose API for writing perf map files

### PR DESCRIPTION
**Pull request by gsallam**
#96123 added support for CPython to write /tmp/perf-<pid>.map files, associating instruction address ranges with a human-readable frame name for the Linux perf profiler.

Two external Python JIT compilers, [Cinder](https://github.com/facebookincubator/cinder/blob/cinder/3.10/Jit/perf_jitdump.cpp#L177-L185) and [Pyston](https://github.com/pyston/pyston/blob/dee48edd457518cf3d2cb93648f9cffe8aad6bce/Python/aot_ceval_jit.c#L5782-L5796), both also independently write to perf map files.

Since perf map files are one-per-process, multiple separate libraries trying to write perf map entries independently can lead to file corruption from simultaneous writes.

It's unlikely for both Cinder and Pyston JITs to be used in the same process, but it's quite reasonable to use one of these JITs along with CPython's native perf trampoline support.

This PR add a C-API to write to the perf map file. It also update the perf trampoline to use the new API to write to the perf map file. 

* Issue: gh-103295

### CODE REVIEW

**Comment by czardoz** [comment]
Looks like the CI is having trouble installing dependencies, I will update this branch with rebase and try again in a bit

**Comment by czardoz** [comment]
I think this is ready for review now. Tagging @pablogsal who is most knowledgeable about this feature (no rush, just letting you know that this is ready for review) :)

**Comment by carljm** [comment]
@gpshead as a reviewer on #96123, would you be willing to take a look at this today? I think it's important that if 3.12 ships with perf-map-writing, which it looks like it will, that there be an API available that makes it possible for third-party extensions to also safely write to the per-process perfmap file without colliding with the built-in perf trampoline feature and causing garbled perfmaps.

File: Doc/c-api/perfmaps.rst [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+   Write one single entry to the ``/tmp/perf-$pid.map`` file. This function is
+   thread safe. Here is what an example entry looks like::
+
+      # address      size  name
+      0x7f3529fcf759 b     py::bar:/run/t.py
+
+   Extensions are encouraged to directly call this API when needed, instead of
+   separately initializing the state by calling :c:func:`_PyOS_PerfMapState_Init`.
+
+.. c:function:: int _PyOS_PerfMapState_Fini(void)
```
Comment (pablogsal) [inline Doc/c-api/perfmaps.rst]: Let's use the PEP 689 nomenclature for these APIs

File: Modules/posixmodule.c [hunk 2]
```diff
@@ -16733,6 +16733,71 @@ INITFUNC(void)
     return PyModuleDef_Init(&posixmodule);
 }
 
+static PerfMapState perf_map_state;
```
Comment (pablogsal) [inline Modules/posixmodule.c]: I don't think this is a good place for this API, given that the other public `trampoline` APIs are in the sys module:

https://docs.python.org/3.12/library/sys.html#sys.activate_stack_trampoline

Let's move this one to the sys module as well.

File: Doc/c-api/perfmaps.rst (hunk 1 above)
Comment (pablogsal) [reply Doc/c-api/perfmaps.rst]: I also would prefer to have ALL of them under the `PyUnstable_` prefix for the time being until we understand more about the usage of the feature.

**Comment by pablogsal** [comment]
Hummm seems that there are a bunch of reference leaks here. I will investigate later

**Comment by czardoz** [comment]
Thanks for reviewing this! The refleak seems to be a side-effect of initializing the trampoline state in the test case. I think the best way to avoid it is to run a subprocess and ensure that the perf map gets created (cc @gsallam)

File: Python/perf_trampoline.c [hunk 3]
```diff
@@ -193,75 +193,28 @@ typedef struct trampoline_api_st trampoline_api_t;
...
-                                  NULL);
-        return;
+    const char *filename = "";
+    if (co->co_filename != NULL) {
+        filename = PyUnicode_AsUTF8(co->co_filename);
     }
-    fprintf(method_file, "%p %x py::%s:%s\n", code_addr, code_size, entry,
-            filename);
-    fflush(method_file);
+    char perf_map_entry[500];
```
Comment (gpshead) [inline Python/perf_trampoline.c]: I tend to not like large things allocated on the stack. When code runs with small stack sizes (as is common in threads started by non-Python C/C++ code in environments I've worked in), this kind of thing can bump into the limit at inopportune times.

I understand why `*_WritePerfMapEntry` wants to take entry_name as a single pre-constructed string for other outside of the CPython runtime uses cases (otherwise I'd suggest passing the utf-8 qualname and filename down directly for formatting in place).

I suggest: malloc and free a temporary buffer around the call to avoid a large buffer on the stack.

File: Doc/c-api/perfmaps.rst
Comment (gpshead) [inline Doc/c-api/perfmaps.rst:1]: It seems worthwhile specifically mentioning that the GIL must be held while calling these APIs.

Otherwise there might be some confusion as they mention that they are thread safe with their own lock, which is true, but I don't want anyone to assume that means the GIL does _not_ apply.

File: Misc/NEWS.d/next/C API/2023-04-14-23-05-52.gh-issue-103295.GRHY1Z.rst
Comment (gpshead) [inline Misc/NEWS.d/next/C API/2023-04-14-23-05-52.gh-issue-103295.GRHY1Z.rst:1]: perhaps add a ReST internal link to the howto/perf_profiling.rst doc?

File: Doc/c-api/perfmaps.rst
Comment (gpshead) [inline Doc/c-api/perfmaps.rst:1]: I suggest linking to this with a one or two sentence mention from the main [howto/perf_profiling.rst doc.](https://github.com/python/cpython/blob/main/Doc/howto/perf_profiling.rst) doc as that'll be the one people are most likely to read first.

**Review by gpshead — CHANGES_REQUESTED** [review]

File: Doc/c-api/perfmaps.rst
Comment (carljm) [reply Doc/c-api/perfmaps.rst:1]: Part of the goal here (and the reason for the separate lock) was to not require the GIL, as e.g. a JIT may want to do multithreaded compilation without holding the GIL, if it can compile without modifying Python objects. (Cinder JIT does this.)

Is there a risk you have in mind that would require the GIL to be held when using this API?

**Comment by carljm** [comment]
(The buildbot failures here were present on main and fixed by https://github.com/python/cpython/pull/104320, thus the merge from main. I'll re-run the buildbots again. Note that the buildbot worker for s390x Fedora buildbots is out of disk space, so those builds will still fail and are failing on other PRs as well.)

**Comment by carljm** [comment]
Hi @gpshead @pablogsal, any further thoughts here?

**Comment by pablogsal** [comment]
>Hi @gpshead @pablogsal, any further thoughts here?

LGTM! Thanks a lot, @carljm and @gsallam for the good work and the patience. 

I will do another buildbot pass, fix the merge conflicts and land this in the weekend. 🚀

File: Include/internal/pycore_ceval.h [hunk 4]
```diff
@@ -66,7 +66,7 @@ typedef struct {
     void (*write_state)(void* state, const void *code_addr,
                         unsigned int code_size, PyCodeObject* code);
     // Callback to free the trampoline state
-    int (*free_state)(void* state);
```
Comment (sunmy2019) [inline Include/internal/pycore_ceval.h]: Why is this change necessary? I see the only usage of this is set to `NULL` as a placeholder, and it is never called. (No code matches the pattern `free_state\(.*` now)


Is it used to support the user-customized function? Buy why removing the `(void* state)` if you need a `void *init_state(void)`?

File: Include/internal/pycore_ceval_state.h [hunk 5]
```diff
@@ -27,7 +27,7 @@ struct trampoline_api_st {
     void* (*init_state)(void);
     void (*write_state)(void* state, const void *code_addr,
                         unsigned int code_size, PyCodeObject* code);
-    int (*free_state)(void* state);
+    void (*free_state)(void);
```
Comment (sunmy2019) [inline Include/internal/pycore_ceval_state.h]: ditto

File: Doc/c-api/perfmaps.rst
Comment (gpshead) [reply Doc/c-api/perfmaps.rst:1]: Upon further code inspection, it appears that only PyUnstable_PerfMapState_Init() requires the GIL (due to the PyErr_Set calls).  Just document that.

Agreed that you don't _want_ a GIL requirement given the purpose, I think the important one is WritePerfMapEntry and that one appears to be, as intended, fine for use w/o the GIL.

**Review by gpshead — APPROVED** [review]
overall LGTM, thanks everybody.  updating the docs to mention the GIL requirement on the _Init method would be good.  and explicitly mentioning that the GIL is _not_ required for WritePerfMapEntry calls seems good as that is one of the main points of the API.

but those are docs changes and can be tweaked later as necesary.

File: Doc/c-api/perfmaps.rst
Comment (carljm) [reply Doc/c-api/perfmaps.rst:1]: Hmm, thanks for highlighting this point! The wrinkle is that `PyUnstable_WritePerfMapEntry` can call `PyUnstable_PerfMapState_Init` (if the perf map state is not already initialized.) We'd rather not force all callers to explicitly call `PerfMapState_Init` themselves.

We could just document that if you are going to call `WritePerfMapEntry` without the GIL (likely a less common case?), then you have to first ensure that `PerfMapState_Init` has already successfully been called in the process (with the GIL).

But I think our preferred option here is to stop setting Python exceptions in `PyUnstable_PerfMapState_Init` entirely, and instead make the entire API lower level and GIL-free. Instead we'd just return one of two error codes from `PerfMapState_Init`: one for "failed to open perf map file" and the other for "failed to allocate lock." If the "failed to open" code is returned, the caller can check `errno` themselves if they need more details.

Any thoughts/concerns about this approach? If it sounds OK, we can update the PR today.

File: Python/perf_trampoline.c [hunk 6]
```diff
@@ -491,10 +449,6 @@ _PyPerfTrampoline_Fini(void)
         tstate->interp->eval_frame = NULL;
     }
     free_code_arenas();
-    if (trampoline_api.state != NULL) {
-        trampoline_api.free_state(trampoline_api.state);
-        trampoline_api.state = NULL;
-    }
```
Comment (sunmy2019) [inline Python/perf_trampoline.c:497]: How about using the same thing as the init?
```c
    if (trampoline_api.state != NULL && trampoline_api.free_state != NULL) {
        trampoline_api.free_state(trampoline_api.state);
        trampoline_api.state = NULL;
    }
```

**Review by sunmy2019 — CHANGES_REQUESTED** [review]

File: Python/perf_trampoline.c (hunk 6 above)
Comment (czardoz) [reply Python/perf_trampoline.c:497]: Keeping in mind that this particular backend for the trampoline API explicitly sets free_state to NULL, I don't quite see a benefit to spending cycles checking it. Can you elaborate on why this is better?

In any case, this will be a no-op, so I don't mind doing this change if there's a benefit to it :)

File: Doc/c-api/perfmaps.rst
Comment (gpshead) [reply Doc/c-api/perfmaps.rst:1]: Returning error codes instead of setting exceptions makes sense to me.

It is still worth documenting that these APIs are intended to have no GIL requirement for others future reference.

File: Python/perf_trampoline.c (hunk 6 above)
Comment (sunmy2019) [reply Python/perf_trampoline.c:497]: If the `init_state` is optionally called, why should we not also call the `free_state`? I don't see why we should have this asymmetry.

If both `init_state` and `free_state` are no-op, maybe we can remove both from `trampoline_api`? 

----
`free_state` is now an unused variable. I think either we should use it, or we should remove it.

Comment (czardoz) [reply Python/perf_trampoline.c:497]: That's an interesting suggestion. Just to make sure I understand correctly - this involves changing the `trampoline_api_st` struct, which is a member of the `_ceval_runtime_state` struct?

Going by the [discussion in the initial implementation](https://github.com/python/cpython/pull/96123#pullrequestreview-1079722071), this API was designed to support different profiling backends, with the `perf` tool being one of them. In the future, I expect we might use this struct to support profiling tools that do require some state initialization and cleanup, for which this is useful to have. 

Changing the `trampoline_api_st` probably warrants a separate discussion. The scope of this particular PR is limited to:
1. Introducing C-APIs for managing the perf-map file.
2. Changing the `perf_trampoline.c` implementation to use the above APIs

Maybe @pablogsal or @gpshead might have more comments on this :)

Comment (czardoz) [reply Python/perf_trampoline.c:497]: Forgot to mention: I share your concern about the asymmetry here, thanks for noticing. We could get rid of the `init_state` call here, since that function is always NULL for the `perf` backend (y)

**Review by gpshead — APPROVED** [review]

File: Python/perf_trampoline.c [hunk 7]
```diff
@@ -193,75 +193,33 @@ typedef struct trampoline_api_st trampoline_api_t;
...
-    const char *filename = PyUnicode_AsUTF8(co->co_filename);
-    if (filename == NULL) {
-        _PyErr_WriteUnraisableMsg("Failed to get filename from code object",
-                                  NULL);
+    const char *filename = "";
+    if (co->co_filename != NULL) {
+        filename = PyUnicode_AsUTF8(co->co_filename);
+    }
+    size_t perf_map_entry_size = snprintf(NULL, 0, "py::%s:%s", entry, filename) + 1;
+    char* perf_map_entry = (char*) malloc(perf_map_entry_size);
```
Comment (sunmy2019) [inline Python/perf_trampoline.c]: Shall we use `PyMem_RawMalloc` or `PyMem_RawFree` here?

File: Python/perf_trampoline.c [hunk 8]
```diff
@@ -193,75 +193,33 @@ typedef struct trampoline_api_st trampoline_api_t;
...
+    if (perf_map_entry == NULL) {
         return;
     }
-    fprintf(method_file, "%" PRIxPTR " %x py::%s:%s\n", (uintptr_t) code_addr, code_size, entry,
-            filename);
-    fflush(method_file);
+    snprintf(perf_map_entry, perf_map_entry_size, "py::%s:%s", entry, filename);
+    PyUnstable_WritePerfMapEntry(code_addr, code_size, perf_map_entry);
+    free(perf_map_entry);
```
Comment (sunmy2019) [inline Python/perf_trampoline.c]: ditto

**Review by sunmy2019 — APPROVED** [review]

**Comment by carljm** [comment]
Refleak buildbot failures are unrelated, they are all failing in `test_mmap`, which is a known issue on main: https://github.com/python/cpython/issues/104698

I think this is ready for merge; @pablogsal I will plan to merge it tomorrow evening unless you have objections!

**Comment by pablogsal** [comment]
> I think this is ready for merge; @pablogsal I will plan to merge it tomorrow evening unless you have objections!

I went ahead and land it myself after another review pass 😉

Great job everyone!
